Make failed download print error

m-ildefons opened the pull request:
https://github.com/rancher/rancher/issues/49788 & https://github.com/rancher/rancher/issues/49790
The machine driver invocation starts by downloading the machine driver binary from Rancher server.
This download may fail due to a variety of reasons. This commit adds an option to curl to print the error message in case of a failed download for easier troubleshooting and debugging.

m-ildefons committed:
Make failed download print error

The machine driver invocation starts by downloading the machine driver
binary from Rancher server.
This download may fail due to a variety of reasons. This commit adds an
option to curl to print the error message in case of a failed download
for easier troubleshooting and debugging.

Signed-off-by: Moritz Röhrich <moritz.rohrich@suse.com>

jakefhyde commented:
@m-ildefons Thanks for the PR, would you mind just sharing an example output of what the message looks like?

m-ildefons commented:
Currently the output may look like this in case of an error:
```
    Failure detected from referenced resource rke-machine.cattle.io/v1,
    Kind=HarvesterMachine with name "test2-pool1-pndcf-vlpb9": Downloading
    driver from
    https://rancher.asan.darkhonor.net/assets/docker-machine-driver-harvester

    Doing /etc/rancher/ssl

    ls: docker-machine-driver-*: No such file or directory

    downloaded file  failed sha256 checksum

    download of driver from
    https://rancher.asan.darkhonor.net/assets/docker-machine-driver-harvester
    failed
```
There is only the hint that the downloaded driver executable is missing and that the checksum of the nonexistent file doesn't match the expected checksum of the actual file>
However it's not immediately clear that the download failed and once you read the source code and realize that the download must have failed, it's not clear what the actual error was.
Moreover it doesn't expose if the error is retriable, if the error is the result of misconfiguration or something else.

In case of custom certificates for example it may be a huge difference whether the certificate chain isn't complete or if the certificate was not signed with the right IP SAN.
But both of these cases would be misconfigurations that the user can correct if they just get informed of what went wrong.

What my goal here is that the output of the script in case of an error with the download command will look something like this:
```
    Failure detected from referenced resource rke-machine.cattle.io/v1,
    Kind=HarvesterMachine with name "test2-pool1-pndcf-vlpb9": Downloading
    driver from
    https://rancher.asan.darkhonor.net/assets/docker-machine-driver-harvester

    Doing /etc/rancher/ssl

    curl: (60) SSL certificate problem: self-signed certificate in certificate chain
    More details here: https://curl.se/docs/sslcerts.html

    curl failed to verify the legitimacy of the server and therefore could not
    establish a secure connection to it. To learn more about this situation and
    how to fix it, please visit the web page mentioned above.

    ls: docker-machine-driver-*: No such file or directory

    downloaded file  failed sha256 checksum

    download of driver from
    https://rancher.asan.darkhonor.net/assets/docker-machine-driver-harvester
    failed
```
This should provide a much clearer picture of what went wrong.

snasovich commented:
@rancher/rancher-security , could you please review if it's a safe change from security perspective?

macedogm commented:
> @rancher/rancher-security , could you please review if it's a safe change from security perspective?

It's a bit hard to say if it will be harmless or not, because we obviously don't know in advance what the error message can eventually be (if it has the possibility of leaking a token passed in the URL or not).

Could we instead output curl's exit code when it's non-zero?

curl provides dozens of different and very specific return codes and they can be used to pinpoint the error instead.

m-ildefons committed:
hide error message

Only put return code in the logs. This makes the code easier to audit
for potential credential leaks.

Signed-off-by: Moritz Röhrich <moritz.rohrich@suse.com>

m-ildefons commented:
@jakefhyde

Now that the actual error message is hidden, an example output would be:

```
    Failure detected from referenced resource rke-machine.cattle.io/v1,
    Kind=HarvesterMachine with name "test2-pool1-pndcf-vlpb9": Downloading
    driver from
    https://rancher.asan.darkhonor.net/assets/docker-machine-driver-harvester

    Doing /etc/rancher/ssl

    Curl failed with error code 60

    ls: docker-machine-driver-*: No such file or directory

    downloaded file  failed sha256 checksum

    download of driver from
    https://rancher.asan.darkhonor.net/assets/docker-machine-driver-harvester
    failed
```

@macedogm please re-review.

m-ildefons commented:
Thanks for the quick reviews. I don't have permissions to merge, please do so for me.
